### PR TITLE
Redesign Dynamo module guards around structural fingerprints

### PR DESCRIPTION
**Pull request by bobrenjc93**
Summary:
- Add a Dynamo nn.Module structural fingerprint guard that snapshots module type, module __dict__ version, structural module dictionaries, hooks, parameters, buffers, and recursive child module structure.
- Normalize large nn.Module guard sets before guard construction so repeated child module identity/type/training/hook/key guards are covered by one root structural guard.
- Add regression coverage for nested-module guard count reduction and recompilation when a child module attribute mutation changes the structural fingerprint.

Root cause problem:
Dynamo emits thousands of repeated leaf guards for deeply nested nn.Module trees, including repeated module identity/type, training, hook, and module-dictionary key guards. The basic nested eager case can therefore spend guard build time and cache metadata on redundant per-leaf checks instead of one normalized structural representation.

Proposed fix:
Introduce an NN_MODULE_STRUCTURE guard backed by a structural fingerprint and run a normalization pass over large nn.Module guard sets. The pass keeps necessary root guards and replaces covered repeated leaf guards with a root-level structural fingerprint guard.

Why this is the right long term fix:
The fingerprint expresses the invariant Dynamo needs at the module-tree boundary: module class, module dictionary shape/version, structural containers, hooks, parameters/buffers/modules, and recursive child structure. This is a more stable guard IR than enumerating the same structural facts thousands of times, while preserving recompilation when the module tree or child module state changes.

Tests:
- USE_NIGHTLY='2.13.0.dev20260427+cpu' python -m pip install --no-build-isolation -v -e .
- PYTHONPATH=/tmp/repos/pytorch/pytorch python -m py_compile torch/_dynamo/guards.py torch/_dynamo/config.py test/dynamo/test_guard_manager.py
- PYTHONPATH=/tmp/repos/pytorch/pytorch python test/dynamo/test_guard_manager.py -k 'NNModuleStructuralGuardTests'
- PYTHONPATH=/tmp/repos/pytorch/pytorch python test/dynamo/test_guard_manager.py

Drafted via Codex, published after manual review by @bobrenjc93

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @azahed98

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@claude /pr-review

**Comment by bobrenjc93** [comment]
@claude /pr-review

**Comment by bobrenjc93** [comment]
@claude /pr-review

**Comment by bobrenjc93** [comment]
@claude /pr-review

**Comment by bobrenjc93** [comment]
[bot-codex-comment] Addressed the latest Claude feedback in 013748730a4: renamed the child-attribute test to make clear it validates preserved leaf value guards, added the requested explanation in the test, and documented that the current structural fingerprint check is a Python tree walk aimed at reducing guard build work/cache metadata until the check can move into the C++ guard manager. I left the current exit-127 CI failures alone because they are broad shard failures and the touched Dynamo guard-manager shard passes.

**Comment by bobrenjc93** [comment]
@claude /pr-review

**Comment by bobrenjc93** [comment]
[bot-codex-comment] Addressed the latest Claude feedback in 7cbbddc4e12: kept the structural accessor tree and added an inline explanation because it is needed for guard-manager dict-tag invalidation to notice child-only module mutations; removed the dead `_replace_output_graph_guards` fallback; represented shared/cyclic module references explicitly so diamond-shaped module graphs are not re-walked; and added a regression test plus a guard-count lower bound. The added shared-module test fails without the implementation change (visit count 2 instead of 1) and passes with it. I left the aarch64 CI failure alone because the touched Dynamo guard-manager shard passed and the job later failed in `test/custom_operator` due to missing `libcustom_ops.so`, unrelated to this Dynamo change.

**Comment by bobrenjc93** [comment]
[bot-codex-comment] Pushed 738195147ad to fix the branch-related export CI failures. Root cause: strict export can construct guards from an `OutputGraphCommon` wrapper that owns persistent `_guards` but has no `tracing_context`; the previous review cleanup removed that fallback. Fix: `_replace_output_graph_guards` now updates `tracing_context.guards_context.dynamo_guards` for normal `OutputGraph` and falls back to updating `output_graph._guards` for `OutputGraphCommon` wrappers. Validated with the failing export reproducer, `NNModuleStructuralGuardTests`, the full `test/dynamo/test_guard_manager.py`, `py_compile`, and Ruff 0.14.4 format/check.

**Comment by bobrenjc93** [comment]
@claude /pr-review

**Comment by bobrenjc93** [comment]
[bot-codex-comment] Addressed the latest Claude feedback in 90054a2edb1: preserved child `NOT_PRESENT_IN_GENERIC_DICT("forward")` guards through structural normalization because the fingerprint only covers nn.Module structural dicts, added a regression that fails without the fix for child `forward` monkey-patching, and added coverage for a real child-module structural mutation. The guard-count test now asserts the structural guard appears, the enabled count is nonzero, and enabled remains lower than the disabled baseline instead of relying on an absolute cap that is no longer valid once forward monkey-patch guards are preserved. Validated the new regression fails without the implementation fix, then passes with it; also ran `NNModuleStructuralGuardTests`, the full `test/dynamo/test_guard_manager.py`, `py_compile`, and Ruff 0.14.4 format/check. I left the aarch64 `libcustom_ops.so` failure and the unstable torchbench `speech_transformer` failure out of this patch because they are not caused by the structural guard correctness issue addressed here.

**Comment by bobrenjc93** [comment]
@claude /pr-review
